### DISCUSSION TITLE
Aggregate MPR lines using Statanalysis

### DISCUSSION THREAD

**Discussion by ghost** [dtcenter/METplus#2440] [Configuration] [answered]
Hi,
I am trying to create a time series of Observation stations using matched pair generated by Pointstat. But I am getting error while doing that, like 

ERROR  :
ERROR  : do_job_aggr_stat() -> the "-line_type" option must be used exactly once and the "-out_line_type" option must be used at least once: -job aggregate_stat -model point_stat -fcst_valid_hour 000000 -fcst_var AOTK -obs_var AOD -line_type MPR
ERROR  :
ERROR  :
ERROR  : main() -> encountered an error value of 1.  Calling clean_up() and usage() before exiting.
ERROR  :

Logfile : /scratch1/NCEPDEV/stmp2/Partha.Bhattacharjee/METplus/Scripts/Regional/logs/metplus.log.20231130062145

Any help on this is appreciated. Thanks.

**Comment by CPKalb** [comment]
Hi:

From your log file, it looks like the job command given to Stat-Analysis is as follows:
-job aggregate_stat -model point_stat -fcst_valid_hour 000000 -fcst_var AOTK -obs_var AOD -line_type MPR

Although the command has the -line_type flag, it is missing the -out_line_type flag, which is the cause for the error.  For any aggregate_stat job, the output line type is different from the input line type and MET needs that information to know what output line type to write and which statistics to compute.  This [table ](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html#id9)shows the options for output line types that can be produced from the MPR line type.  What type of output statistics are you trying to create for the time series?

If you don't want to compute statistics, but just want to separate out certain variables or lines to create a time series, then the [filter](https://met.readthedocs.io/en/latest/Users_Guide/stat-analysis.html#filter-stat-lines) job would be the one to use.

Christina

**Reply by ghost** [reply]
Thanks Christina. I put STAT_ANALYSIS_JOB1 = -job filter -dump_row [dump_row_file] but it still shows 0 size in the output. Config resides here : /scratch1/NCEPDEV/stmp2/Partha.Bhattacharjee/METplus/Scripts/Regional/StatAnalysis_Pointstat.conf  and logfile : /scratch1/NCEPDEV/stmp2/Partha.Bhattacharjee/METplus/Scripts/logs.

Thanks.

**Reply by CPKalb** [reply]
I am not able to see the log directory listed above so could not take a look at your log file.

However, when I run the configuration file at /scratch1/NCEPDEV/stmp2/Partha.Bhattacharjee/METplus/Scripts/Regional/StatAnalysis_Pointstat.conf, I get the following information:
DEBUG 2: STAT Lines read     = 16594
DEBUG 2: STAT Lines retained = 0
DEBUG 2:
DEBUG 2: Processing Job 1: -job filter -model point_stat -fcst_valid_hour 000000 -fcst_var AOTK -obs_var AOD -line_type MPR -dump_row ./point_stat_000000L_20220508_000000V.stat

From this information, there are 0 stat lines that are being output to the file.  What this means is that all the data is being filtered out due to the settings in the job command.  The issue here is your model setting in the configuration.  The MODEL1 variable in your configuration is set to point_stat.  However, the model in your output from Point-Stat is RRFS.  So, MODEL1 needs to be changed to:
MODEL1 = RRFS

Making that change, I'm able to get multiple lines of output from Stat-Analysis.  Does this work on your end?

Christina

**Reply by ghost** [reply]
Thanks Christina, it helped to resolve the issue. 
However, looking into pointstat output, I do have another question on using Statanalysis part. So, in .stat file from pointstat, it matches model fcst of a particular valid hour against number of station observations around that particular hour (using time window of 30+/- mins around valid time). I am wondering whether it is possible to take an average of each of the stations observation on that time window using Stat analysis and write out one value for each stations for both model fcst and obs for that particular valid hour. Thanks.

**Reply by CPKalb** [reply]
Hi:

It is possible to take an averages of the station observations.  In this case, the summary option in Stat-Analysis is the one that computes averages.   It will also compute additional information besides the mean, including the standard deviation, percentiles, the max, interquartile range, and more.

Here, you would need 2 jobs, one to average the forecast variable and another to average the observation variables.  The configuration would look something like the following:
STAT_ANALYSIS_JOB1 = -job summary -line_type MPR -column FCST_VAR
STAT_ANALYSIS_JOB2 = -job summary -line_type MPR -column OBS_VAR

Christina

**Reply by ghost** [reply]
Thanks Christina. When I add those two lines in config, I am seeing this in job.out (no averaging of individual stations at each forecast hr, that I would like to see) : 

JOB_LIST:      -job summary -model RRFS -fcst_valid_hour 230000 -fcst_var AOTK -obs_var AOD -line_type MPR -column FCST_VAR -out_alpha 0.05000
COL_NAME: LINE_TYPE COLUMN   TOTAL MEAN MEAN_NCL MEAN_NCU MEAN_BCL MEAN_BCU STDEV STDEV_BCL STDEV_BCU MIN P10 P25 P50 P75 P90 MAX IQR RANGE WMO_TYPE WMO_MEAN WMO_WEIGHTED_MEAN
 SUMMARY: MPR       FCST_VAR  1309    0        0        0       NA       NA     0        NA        NA   0   0   0   0   0   0   0   0     0     MEAN        0                 0

JOB_LIST:      -job summary -model RRFS -fcst_valid_hour 230000 -fcst_var AOTK -obs_var AOD -line_type MPR -column OBS_VAR -out_alpha 0.05000
COL_NAME: LINE_TYPE COLUMN  TOTAL MEAN MEAN_NCL MEAN_NCU MEAN_BCL MEAN_BCU STDEV STDEV_BCL STDEV_BCU MIN P10 P25 P50 P75 P90 MAX IQR RANGE WMO_TYPE WMO_MEAN WMO_WEIGHTED_MEAN
 SUMMARY: MPR       OBS_VAR  1309    0        0        0       NA       NA     0        NA        NA   0   0   0   0   0   0   0   0     0     MEAN        0                 0

/scratch1/NCEPDEV/stmp2/Partha.Bhattacharjee/METplus/Scripts/Regional/logs

**Reply by CPKalb** [reply]
Sorry, the job columns should just say FCST and OBS rather than what I originally gave you.  So, if I amend the job commands to:
STAT_ANALYSIS_JOB1 = -job summary -line_type MPR -column FCST
STAT_ANALYSIS_JOB2 = -job summary -line_type MPR -column OBS

Then I am able to get output that is not all zeroes.

Christina

**Reply by CPKalb** [reply]
Another thing to note here is that your STAT_ANALYSIS_OUTPUT_TEMPLATE is set to a static value of job.out.  This means you will only get one output file, job.out, and it will only contain data from the last Stat-Analysis run (the last run will overwrite the data from all the previous runs).  If you want to save data from each Stat-Analysis run, then you will need to add time template information to your output file name.  

Stat-Analysis doesn't use time template information in the same way that other METplus wrappers do.  This is because when aggregating in time, the concept of a single init or valid time no longer makes sense.  Instead, Stat-Analysis takes a number of [special keys](https://metplus.readthedocs.io/en/latest/Users_Guide/wrappers.html#additional-filename-template-tags).  As an example, to add the valid beg time including the hour, the template would look like the following:

STAT_ANALYSIS_OUTPUT_TEMPLATE = job_{fcst_valid_beg?fmt=%Y%m%d}_{fcst_valid_hour?fmt=%HH}.out

Christina

**Reply by CPKalb** [reply]
Hi:

Were you able to get this up and running as expected?

Thanks,
Christina

**Reply by ghost** [reply]
Thanks @CPKalb , yes it worked and generated job.out at different valid times. 
However, I am looking to aggregate data little differently in Stat Analysis, not like taking mean of all stations and model forecasts at each valid time together to get a single value in job.out . If you check on pointstat output and open any .stat file (for example  pointstat/point_stat_170000L_20200901_170000V.stat), you will notice for each model forecasts over each of stations, there is one model forecast value of that particular hour, but different point observations around that hour (as I am using a 30 min window to allow more observations around particular model forecast hour). All I would like to do in Statanalysis is to average all those station observations for that forecast hour to have one average value like the model forecast and ultimately write out values of all stations and matching model forecasts in the out file. Hope I have explained it better now. 
Many thanks.

**Reply by CPKalb** [reply]
Hi:

It sounds like you are wanting to average all the observations for a particular forecast hour, rather than having the same model matched to observations at slightly different times around the hour.  This is best accomplished in your Point-Stat run, rather than using Stat-Analysis.  Time averaging in Point-Stat is done using the [OBS_SUMMARY](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#obs-summary) option.  For a simple average, this would be set to UW_MEAN.  

OBS_SUMMARY is not yet an option yet in METplus for Point-Stat, so the way to set this variable is using a config override.  Specifically, you would want to set the following in your Point-Stat METplus configuration file:
POINT_STAT_MET_CONFIG_OVERRIDES = obs_summary = UW_MEAN;

Then, run Point-Stat, and you should get only one output each hour matched to an average observation.

Christina

**Reply by georgemccabe** [reply]
Note: Starting in METplus v6.0.0 (upcoming release), `obs_summary` can be set with `POINT_STAT_OBS_SUMMARY = UW_MEAN` instead of `POINT_STAT_MET_CONFIG_OVERRIDES`.

**Reply by ghost** [reply]
I have set POINT_STAT_OBS_SUMMARY = UW_MEAN in the config but getting similar result as before, that is without doing the averages of station obs around model forecast time. Check this :  point_stat_140000L_20200906_140000V.stat in the output. Do I need to set something more in the config.

**Reply by CPKalb** [reply, marked as answer]
POINT_STAT_OBS_SUMMARY will be available when METplus version 6.0 comes out, but we we are currently on METplus version 5.1.  So, at this time POINT_STAT_OBS_SUMMARY does not work.  For now, you will want to set this variable using a config override.  So, remove the POINT_STAT_OBS_SUMMARY = UW_MEAN in your configuration and add this line instead:
POINT_STAT_MET_CONFIG_OVERRIDES = obs_summary = UW_MEAN;

This produces only one output per station every hour when I run it on my end.  Does it work on your end?

Thanks,
Christina

**Reply by ghost** [reply]
Yes, it is working fine. Thanks Christina for all your help and patience.

**Reply by CPKalb** [reply]
Great.  I am going to lock this discussion to prevent future posts.  If you have additional information, feel free to start a new discussion.
